src/VisualStudio/Core/Def/Telemetry/VisualStudioTelemetryInitializer.cs

```diff
@@ -0,0 +1,89 @@
+{
+    [ExportEventListener(WellKnownEventListeners.Workspace, WorkspaceKind.Host), Shared]
+    internal class VisualStudioTelemetryInitializer : IEventListener<object>
+    {
+        private readonly VisualStudioWorkspace _visualStudioWorkspace;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IAsynchronousOperationListener _asynchronousOperationListener;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public VisualStudioTelemetryInitializer(
+            VisualStudioWorkspace visualStudioWorkspace,
+            IThreadingContext threadingContext,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _visualStudioWorkspace = visualStudioWorkspace;
+            _threadingContext = threadingContext;
+            _asynchronousOperationListener = listenerProvider.GetListener(FeatureAttribute.Telemetry);
+        }
+
+        public void StartListening(Workspace workspace, object serviceOpt)
+        {
+            if (workspace is VisualStudioWorkspace)
+            {
+                _ = StartListeningAsync();
+            }
+        }
+
+        private async Task StartListeningAsync()
+        {
+            // Have to catch all exceptions coming through here as this is called from a
+            // fire-and-forget method and we want to make sure nothing leaks out.
+            try
+            {
+                using var token = _asynchronousOperationListener.BeginAsyncOperation(nameof(InitializeTelemetryAsync));
+                await InitializeTelemetryAsync(_threadingContext.DisposalToken).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                // Cancellation is normal (during VS closing).  Just ignore.
+            }
+            catch (Exception e) when (FatalError.ReportAndCatch(e))
+            {
+                // Otherwise report a watson for any other exception.  Don't bring down VS.  This is
+                // a BG service we don't want impacting the user experience.
+            }
+        }
+
+        private async Task InitializeTelemetryAsync(CancellationToken cancellationToken)
+        {
+            // Fetch the session synchronously on the UI thread; if this doesn't happen before we try using this on
+            // the background thread then we will experience hangs like we see in this bug:
+            // https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?_a=edit&id=190808 or
+            // https://devdiv.visualstudio.com/DevDiv/_workitems?id=296981&_a=edit
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            var session = TelemetryService.DefaultSession;
+
+            // Now that the telemetry session has been initialized, the rest of the work can happen off of the UI thread.
+            await TaskScheduler.Default;
```

Comment:
❓ Since the initialization is performed in a fire-and-forget manner, how do we ensure dependent operations block on its completion?

Reply:
➡️ Rather than ensure initialization completes before telemetry events are fired, this code simply _ignores_ events prior to initialization. This is the same behavior we had prior to this change, so it seems fine.